build.rs

```diff
@@ -1,10 +1,116 @@
-use std::env;
+use regex::Regex;
+use std::fmt::{Display, Formatter};
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use std::{env, fmt};
+
+#[derive(Clone, Debug)]
+pub struct Target {
+    pub architecture: String,
+    pub vendor: String,
+    pub system: Option<String>,
+    pub abi: Option<String>,
+}
+
+impl Target {
+    pub fn as_strs(&self) -> (&str, &str, Option<&str>, Option<&str>) {
+        (
+            self.architecture.as_str(),
+            self.vendor.as_str(),
+            self.system.as_deref(),
+            self.abi.as_deref(),
+        )
+    }
+}
+
+impl Display for Target {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "{}-{}", &self.architecture, &self.vendor)?;
+
+        if let Some(ref system) = self.system {
+            write!(f, "-{}", system)
+        } else {
+            Ok(())
+        }?;
+
+        if let Some(ref abi) = self.abi {
+            write!(f, "-{}", abi)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub fn ndk() -> String {
+    env::var("ANDROID_NDK").expect("ANDROID_NDK variable not set")
+}
+
+pub fn target_arch(arch: &str) -> &str {
+    match arch {
+        "armv7" => "arm",
+        "aarch64" => "arm64",
+        "i686" => "x86",
+        arch => arch,
+    }
+}
+
+fn host_tag() -> String {
+    // Because this is part of build.rs, the target_os is actually the host system
+    if cfg!(target_os = "windows") {
+        "windows-x86_64"
+    } else if cfg!(target_os = "linux") {
+        "linux-x86_64"
+    } else if cfg!(target_os = "macos") {
+        "darwin-x86_64"
+    } else {
+        panic!("host os is not supported")
+    }
+    .to_string()
+}
+
+/// Get NDK major version from source.properties
+fn ndk_major_version(ndk_dir: &Path) -> u32 {
+    // Capture version from the line with Pkg.Revision
+    let re = Regex::new(r"Pkg.Revision = (\d+)\.(\d+)\.(\d+)").unwrap();
```

Comment:
this is just simple enough that I wonder if you need to bring in a full regex implementation here, that being said, it is easier, but if you did the parsing yourself, i bet compile times would be faster. hard to say, to be honest, but I thought I'd mention it.